app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -113,25 +135,43 @@ public void onClickByLocation(View view) {
                         .setLocation(location)
                         .build();
-                weatherService.getCurrentWeatherInfo(weatherRequest, new WeatherService.WeatherServiceCallback() {
-                    @Override
-                    public void onWeatherInfoObtained(@NonNull WeatherInfo weatherInfo) {
-                        showWeatherView.setText(getString(R.string.template_weather_message, weatherInfo.getTemperature(), weatherInfo.getVisibility(), weatherInfo.getHumidity(), weatherInfo.getWindSpeed()));
-                        weatherByLocationButton.setEnabled(true);
-                    }
-
-                    @Override
-                    public void onError() {
-                        showWeatherView.setText(getText(R.string.default_weather_message).toString());
-                        Toast.makeText(MainActivity.this, getText(R.string.error_wrong_request).toString(), Toast.LENGTH_SHORT).show();
-                        weatherByLocationButton.setEnabled(true);
-                    }
-                });
+                try {
+                    weatherService.getCurrentWeatherInfo(weatherRequest, new IWeatherServiceCallback.Stub() {
+                        @Override
+                        public void onWeatherInfoObtained(@NonNull WeatherInfo weatherInfo) {
+                            runOnUiThread(new Runnable() {
+                                @Override
+                                public void run() {
+                                    showWeatherView.setText(getString(R.string.template_weather_message, weatherInfo.getTemperature(), weatherInfo.getVisibility(), weatherInfo.getHumidity(), weatherInfo.getWindSpeed()));
+                                    editCityView.setText("");
+                                    weatherByLocationButton.setEnabled(true);
+                                }
+                            });
+                        }
+
+                        @Override
+                        public void onError() {
+                            runOnUiThread(new Runnable() {
+                                @Override
+                                public void run() {
+                                    showWeatherView.setText(getText(R.string.default_weather_message).toString());
+                                    Toast.makeText(MainActivity.this, getText(R.string.error_wrong_request).toString(), Toast.LENGTH_SHORT).show();
+                                    editCityView.setText("");
+                                    weatherByLocationButton.setEnabled(true);
+                                }
+                            });
+                            Log.i(TAG, "onClickByLocation: onError");
+                        }
+                    });
+                } catch (RemoteException e) {
+                    e.printStackTrace();
+                    Log.i(TAG, "onClickByLocation: RemoteException during weather service request");
+                }
```

Comment:
Try to define single IWeatherServiceCallback.Stub as class field and use it while calling getCurrentWeather().